packager/packager.cc

```diff
@@ -0,0 +1,731 @@
+                    {core::AutocorrectSuggestion::Edit{core::Loc(ctx.file, nameLoc), replacement}}});
+            }
+        }
+        constLit = ast::cast_tree<ast::UnresolvedConstantLit>(constLit->scope);
+    }
+}
+
+// Gets the package name in `tree` if applicable.
+PackageName getPackageName(core::MutableContext ctx, ast::UnresolvedConstantLit &constantLit) {
+    PackageName pName;
+    pName.loc = constantLit.loc;
+
+    {
+        auto current = &constantLit;
+        while (current != nullptr) {
+            pName.fullName.emplace_back(current->cnst);
+            current = ast::cast_tree<ast::UnresolvedConstantLit>(current->scope);
+        }
+    }
+    reverse(pName.fullName.begin(), pName.fullName.end());
+    ENFORCE(!pName.fullName.empty());
+
+    // Foo::Bar => Foo_Bar_Package
+    auto mangledName = absl::StrCat(absl::StrJoin(pName.fullName, "_", NameFormatter(ctx)), "_Package");
+    pName.mangledName = ctx.state.enterNameConstant(mangledName);
+
+    return pName;
+}
+
+bool isReferenceToPackageSpec(core::Context ctx, ast::TreePtr &expr) {
+    auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    return constLit != nullptr && constLit->cnst == core::Names::Constants::PackageSpec();
+}
+
+unique_ptr<ast::UnresolvedConstantLit> copyConstantLit(ast::UnresolvedConstantLit *lit) {
+    auto copy = lit->deepCopy();
+    // Cast from Expression to UnresolvedConstantLit.
+    auto copyUcl = ast::cast_tree<ast::UnresolvedConstantLit>(copy);
+    ENFORCE(copyUcl != nullptr);
+    unique_ptr<ast::UnresolvedConstantLit> rv(copyUcl);
+    copy.release();
+    return rv;
+}
+
+ast::TreePtr name2Expr(core::NameRef name, ast::TreePtr scope = ast::MK::EmptyTree()) {
+    return ast::MK::UnresolvedConstant(core::LocOffsets::none(), move(scope), name);
+}
+
+ast::TreePtr PackageName::getFullNameLiteral(core::LocOffsets loc) const {
+    ast::TreePtr name = ast::MK::EmptyTree();
+    for (auto namePart : fullName) {
+        name = name2Expr(namePart, move(name));
+    }
+    // Outer name should have the provided loc.
+    if (auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(name)) {
+        name = ast::MK::UnresolvedConstant(loc, move(lit->scope), lit->cnst);
+    }
+    return name;
+}
+
+ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameRef fun, ast::TreePtr &expr) {
+    auto target = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    if (target == nullptr) {
+        if (auto e = ctx.beginError(expr->loc, core::errors::Packager::InvalidImportOrExport)) {
+            e.setHeader("Argument to `{}` must be a constant", fun.show(ctx));
+        }
+    }
+    return target;
+}
+
+struct PackageInfoFinder {
+    unique_ptr<PackageInfo> info = nullptr;
+    vector<unique_ptr<ast::UnresolvedConstantLit>> exported;
+    vector<unique_ptr<ast::UnresolvedConstantLit>> exportedMethods;
```

Comment:
To echo my other point: I think you're going to run into the same LocOffset problems with `exportedMethods` (if you hadn't already started adapting the changes to it too).

I think the fun thing about this change is that we should be able to delete `copyConstantLit`, which I thought was a bit janky from the get go (anything that uses `.release()` is a hack).

Reply:
`exportedMethods` aren't embedded in other files, though -- so they will be fine.

Reply:
I removed `copyConstantLit` and introduced a `FullyQualifiedName` class that holds the vector of names.